src/panel-ladder.tsx

```diff
@@ -0,0 +1,182 @@
+/**
+ * Ladder Panel
+ *
+ * Panel for ladder formats and associated ladder tables.
+ *
+ * @author Adam Tran <aviettran@gmail.com>
+ * @license AGPLv3
+ */
+
+class LadderRoom extends PSRoom {
+	readonly classType: string = 'ladder';
+	readonly notice?: string;
+	ladderData?: string;
+	selectedFormat?: string;
+
+	constructor(options: RoomOptions) {
+		super(options);
+	}
+	setLadderData = (ladderData: string | undefined) => {
+		this.ladderData = ladderData;
+		this.update(null);
+	};
+	setFormat = (selectedFormat: string | undefined) => {
+		const { teams } = PS;
+		this.selectedFormat = selectedFormat;
+		if (selectedFormat !== undefined) {
+			if (teams.usesLocalLadder) {
+				this.setLadderData(undefined); // Loading...
+				this.send(`/cmd laddertop ${selectedFormat} ${toID(this.searchValue)}`);
+			} else {
+				this.requestLadderData();
+			}
+		} else {
+			this.update(null);
+		}
+	};
+	requestLadderData = (searchValue?: string) => {
+		this.setLadderData(undefined); // "Loading..."
+		$.get('/ladder.php', { // TO REVIEW: I imagine this may need to be changed to use PSLoginServer, but that does not have support for prepping GETs with query parameters
+			format: this.selectedFormat,
+			server: Config.server.id.split(':')[0],
+			output: 'html',
+			prefix: toID(searchValue),
+		},
+		this.setLadderData);
+	};
+}
+
+interface LadderPanelState {
+	showHelp: boolean;
+	searchValue: string;
+	lastSearch: string;
+}
+
+class LadderPanel extends PSRoomPanel<LadderRoom, LadderPanelState> {
+	constructor() {
+		super();
+		this.state = { showHelp: false, searchValue: '', lastSearch: '' };
+	}
+	subscriptions: PSSubscription[] = [];
+	componentDidMount = () => {
+		const { room } = this.props;
+		this.subscriptions.push(room.subscribe((response: any) => {
+			if (response) {
+				const [ format, ladderData ] = response;
+				if (room.selectedFormat === format) {
+					room.ladderData = ladderData || '<p>Error getting ladder data from server</p>';
+				}
+			}
+			this.forceUpdate();
+		}));
+		this.subscriptions.push(PS.teams.subscribe(() => {
+			this.forceUpdate();
+		}));
+	};
+	componentWillUnmount() {
+		for (const subscription of this.subscriptions) {
+			subscription.unsubscribe();
+		}
+		this.subscriptions = [];
+	}
+	setShowHelp = (showHelp: boolean) => () => this.setState({ showHelp });
+	setSearchValue = (e: Event) => this.setState({ searchValue: (e.currentTarget as HTMLInputElement).value });
```

Comment:
My plan for the Preact client (which I guess I forgot to document) is for button dispatch to work through commands, so the button would look like `<button name="cmd" value="/searchvalue [number]">` and then the command would be dispatched in `handleMessage`.

Not really a big deal for ladder, but probably a useful thing to remember in general. My plan is that mousing over buttons would display the command in the closest text box, which would be a nice hint for the user in some cases.

Reply:
Marking resolved (discussed in Discord)